### DISCUSSION TITLE
Registry Metadata

### DISCUSSION THREAD

**Discussion by esoterra** [bytecodealliance/registry#116] [General]
There are two related questions about registry metadata we need to answer:

1. What kind of metadata do we need to store for packages and/or releases? (e.g. author, license, website, repo)
2. What format should metadata be encoded in? (e.g. JSON, proto)

**Comment by esoterra** [comment]
In my opinion, the format/encoding we choose needs to
1. be able to support common types like strings, numbers, dates, etc.,
2. be able to be extended easily by adding extra fields or attributes, and
3. be easily encoded/decoded on a variety of platforms and languages.

JSON and protobuf can both meet these requirements.
* JSON advantages: easy to read, easy to write, tons of support, does not require codegen or out-of-band schema.
* Protobuf advantage: more compact, represents byte data without `base64`.

**Reply by kriswuollett** [reply]
If metadata about some content needs to be exposed in an API in addition to being embedded within a container format, I'd still lean towards my general opinion that it would be nice if metadata format is defined in `proto3` and get the [official JSON encoding ](https://protobuf.dev/programming-guides/proto3/#json)for free out of it. IMO it should be preferred to minimize writing boilerplate than it is to have avoid having a less normally seen JSON structure style.

If there is, perhaps eventually, an API call that searches for packages, the metadata can be encoded directly for REST or gRPC endpoints within the requests/responses.

However, if/when the metadata would be encoded into a file format that is to be loaded by some WebAssembly runtime/platform library, then I'd guess it should be the JSON encoding to avoid requiring developers from pulling in Protobuf dependencies if they don't want to.

**Reply by esoterra** [reply]
I think it's important to distinguish here that currently our use of protobuf is constrained purely to the encoding of cryptographic data structures and logs. Everything else is done via a REST API and through data encoded in JSON and people (I believe including @peterhuene) have a genuine interest in designing that API and those JSON documents to be idiomatic and not derived from `proto3` in ways that tend not to be.

Proposing that we move to be `proto`-first in all cases is a *big* discussion that is far from having the consensus needed to be presumed here.

Another point here is that we should be prepared to use the encoding we pick here in *all* contexts because it's not generally possible to convert back and forth between JSON and protobuf or to decode/re-encode the same format in a way that preserves the hash of the data.

**Reply by lann** [reply]
Indeed pbjson is pretty awkward to work with if you aren't already using protobuf generation. I wouldn't use proto for schema definition unless proto encoding was a priority.

**Reply by kriswuollett** [reply]
Just to note, the serialization of Protobuf is [not guaranteed to be stable](https://protobuf.dev/programming-guides/encoding/#implications). Flatbuffers, while possibly more interesting since parsing isn't required, [its encoding is also not guaranteed to be stable](https://groups.google.com/g/flatbuffers/c/v2RkM3KB1Qw). Currently using something like fields alphabetically ordered, null fields omitted JSON encoding for things like cryptographic proofs? No point was I suggesting to sign encoded Protobuf data.

**Reply by lann** [reply]
One of the main reasons we're using protobuf for the signed payloads is that we are storing them as encoded by the user to sidestep any encoding canonicalization problems wrt signing.

**Reply by kriswuollett** [reply]
Even for an end-user, IMO the serialization should be stable; otherwise, if they update their toolchain, they might not be able to reproduce a use case with the same output given some content and metadata. Similarly, is there a use case to detect or search for duplicate content or metadata in the warg system? If so, stable JSON where needed is great.

Additionally, as much as I like using Protobufs, I'd think one wouldn't want to work with them if they are creating a warg client/server in WASM to be run in a place like a browser, since it would require pulling in additional dependency and complicate the build process.

**Reply by lann** [reply]
> Even for an end-user, IMO the serialization should be stable; otherwise, if they update their toolchain, they might not be able to reproduce a use case with the same output given some content and metadata.

I don't understand the point here. Could you give an example?

**Reply by kriswuollett** [reply]
I do not have a specific example because I am not familiar with the contents of the signed payload that is referred to in https://github.com/bytecodealliance/registry/discussions/116#discussioncomment-5919131. But in general examples would be nearly any case where it is structured data matching a schema owned by warg, for the API, etc., on a serialized version of which a hash is applied. Content that this does not need to apply to is opaque content like compiled code, which may be pointed to by a hash itself instead of being included directly.

Just suggesting something to look into if it affects tests/logs reproducibility due to something like upgrading/changing Protobuf implementation, e.g., integration tests would break if checking for the hash in them.

**Reply by lann** [reply]
The signed payloads are part of the registry log. Based on the discussion in the [other thread](https://github.com/bytecodealliance/registry/discussions/116#discussioncomment-5917783), it doesn't sound like we'll be storing package metadata there directly.

It's true that stable encoding is convenient, but I think at this point what we need to decide is whether a _particular_ stable encoding should be required by the protocol and enforced by the server. Certainly any client would be free to use a stable e.g. JSON encoder.

**Comment by lann** [comment]
Where will the metadata be stored?

**Reply by esoterra** [reply]
In my view there are two requirements which dictate where metadata can/should be stored.
1. Metadata should be verifiable, which means its hash must be included directly or transitively in the release entry.
2. Metadata contains arbitrary user-controlled data (with the potential to include PII) therefore it cannot exist directly in the log itself.

So, I think as a result metadata should be stored separately from the logs as content with its hash included in the log.
This could be stored using the existing content storage mechanisms and/or in a different way.

**Reply by lann** [reply]
I probably wouldn't use protobuf unless the metadata is itself in the release log. JSON seems like a reasonable choice.

**Reply by esoterra** [reply]
Do you have any thoughts on the larger REST/JSON-first vs. gRPC/proto-first discussion?

**Reply by silesmo** [reply]
> In my view there are two requirements which dictate where metadata can/should be stored.
> 
>     1. Metadata should be verifiable, which means its hash must be included directly or transitively in the release entry.
> 
>     2. Metadata contains arbitrary user-controlled data (with the potential to include PII) therefore it cannot exist directly in the log itself.
> 
> 
> So, I think as a result metadata should be stored separately from the logs as content with its hash included in the log. This could be stored using the existing content storage mechanisms and/or in a different way.

How will this work in practice if you make an edit to a components description in the metadata. Would this impact the resolution of dependencies somehow if you have the same component included in different modules in the component that just have a diff in the metadata? Would the resolution in that case think of them as two different components?

**Reply by silesmo** [reply]
If metadata is verifiable do you mean all the defined fields are verifiable? Or all the defined and custom fields? @Kylebrown9

**Reply by lann** [reply]
Dependency information will be embedded in components. I don't know of any dependency metadata that will need to be external.

**Reply by esoterra** [reply]
There are some examples in [the other thread](https://github.com/bytecodealliance/registry/discussions/116#discussioncomment-5920217) of how different metadata is handled. In general, I also don't expect you to be able to "edit" metadata of prior releases. To my knowledge that's not possible with existing systems (e.g. cargo) anyway. If by edit you just mean to have a different value in the metadata than in the component the general approach is to avoid having the same info in two places.

The idea here is that the metadata as a whole (so including all fields) would be hashed and that hash included in the verifiable data structures.

**Comment by esoterra** [comment]
In terms of the fields we need to store, it would make sense to include some common subset of typical `Cargo.toml`, `package.json`, etc. metadata at the top level for all packages and allow for other fields to be added ad-hoc.

This suggests some questions to explore like
1. What existing systems (e.g. `cargo`, `npm`) metadata should we consider?
2. What data is common to all or most of them?
3. What data is necessary/useful to have about all packages globally when looking at an applications dependencies? e.g. licenses

**Reply by Scroul** [reply]
I think these should be most of the fields that exist in most different package registry system and I think all of them in some form can be nice to support. Tried to include the common formats used by the fields.
* name - string
* version - string
* authors/contributors - string[]
* description - string
* documentation - URL
* homepage - URL
* code repo - URL
* license/licenses/license-files - string, string[]
* funding/donations - string[], object[]
* keywords/categories - string[]

Dependencies is of course a standard field more of a question regarding if we want one dependencies field or if we want to give a selection of different dependency fields.
* dependencies/dev-dependencies/build-dependencies

Not sure if this is something we would want as part of a standard or if it is better to leave to tools/registries. 
* scripts/commands

I also belive that a runtime field could be nice to have. As an array of objects with runtime version and runtime name. Can give a good way to display what something has been test and is known to work with. Also a custom/extra section for tools/registries to store data they want to store that is not part of any of the standard fields.

**Reply by lann** [reply]
> version - string

The component model currently includes a semver-like major/minor interface version in component import declarations. Should we enforce semver for the version field?

**Reply by Scroul** [reply]
Yeah, I think it is a good idea to do that to match what is done in the component model.

**Reply by peterhuene** [reply]
The release's (semver-compatible) version is already encoded in the log and is authoritative, so it might be redundant to also include it in the metadata.

**Reply by lann** [reply]
> license

Should we define/enforce this to be a [SPDX license](https://spdx.org/licenses/)?

**Reply by esoterra** [reply]
To summarize the above and add a few new questions, here's my thoughts on each of the above.
* name - do not duplicate in metadata, already constrained/defined by the verifiable log and map
* version - do not duplicate in metadata, already included in the package log release entry
* authors, description, documentation, homepage, code repo - all good/easy inclusions validating URLs would be nice
* license - not too familiar with SPDX but happy to use some standardized identifiers
* funding/donations - not clear on what this means, do you have examples/ideas?
* keywords/categories - sounds good
* dependencies - already encoded in component binary as imports with identifiers
* scripts/commands - I don't think we need this

**Reply by Scroul** [reply]
Regarding funding/donations I was mainly thinking about something like https://docs.npmjs.com/cli/v6/configuring-npm/package-json#funding 

Would possible be something for open source projects to use give a good way for them show where support can be given to them. Would in some ways be covered by other fields since information like that often exist in readmes, docs and  homepage. So would be if we want to display something like that. 

Tldr: URL or a list URLs

**Reply by lann** [reply]
Would it make sense to have a `links: {<type>: <url>}` map with some conventional keys (`"homepage"`, `"docs"`, `"repo"`, `"funding"`)? An interface could choose to display or throw away any keys it doesn't understand.

(or `links: {<type>: [<url>]}` or `links: [{type: <type>, url: <url>}, ...]`)

**Reply by esoterra** [reply]
I'm open to something like that. If we do that, we should pick standard names for common fields like the examples you gave so people don't inconsistently use e.g. `"repo"`/`"repository"`, `"docs"`/`"documentation"`, ...

**Reply by Scroul** [reply]
I think a runtimes filed could be nice to have ex `"runtimes": { "wasmtime": " 8.0.0" }`. 

I see it as a this is known to run as expected in the runtimes listed here at the versions listed here.

**Reply by esoterra** [reply]
I think if we need to mark our components with what versions of what runtimes can run them, other parts of the component model design have failed. You should be able to tell whether you can run a component entirely based on its interface/world and the version of the component model it specifies in its binary. As runtimes become able to support the component model and new versions of it those runtimes will become able to run components they didn't use to.

i.e. Runtimes should say "I support component-model preview2 and the http-proxy world" instead of packages saying "I support wasmtime"

**Reply by esoterra** [reply]
There are also going to be many host platforms e.g. compute@edge (Fastly), spin platform (Fermyon), ... that all use Wasmtime and could (but hopefully won't) have functional differences in the host-bindings/implementation they provide to components. These kinds of host platform differences are probably more likely to accidentally introduce some differences (e.g. one preserves header order the other doesn't) that cause components to run differently than the runtimes they use internally are. If we have a bunch of runtimes and some of them can't correctly run components that's a clearer/bigger issue.

**Reply by silesmo** [reply]
> To summarize the above and add a few new questions, here's my thoughts on each of the above.
> 
>     * name - do not duplicate in metadata, already constrained/defined by the verifiable log and map
> 
>     * version - do not duplicate in metadata, already included in the package log release entry
> 
>     * authors, description, documentation, homepage, code repo - all good/easy inclusions validating URLs would be nice
> 
>     * license - not too familiar with SPDX but happy to use some standardized identifiers
> 
>     * funding/donations - not clear on what this means, do you have examples/ideas?
> 
>     * keywords/categories - sounds good
> 
>     * dependencies - already encoded in component binary as imports with identifiers
> 
>     * scripts/commands - I don't think we need this

When we get to adding search in the future how would this work in practice? I would have assumed the search would be based on the metadata to some degree and if information like the name isn't stored there, then we would need to fetch information from multiple sources, then stitch it back together? Wouldn't you in practice copy that information to the metadata store when you upload a package anyway?

**Comment by lann** [comment]
Cargo's approach to license metadata seems good to me: https://doc.rust-lang.org/cargo/reference/manifest.html#the-license-and-license-file-fields

> Note: crates.io requires either license or license-file to be set.

**Reply by esoterra** [reply]
In today's meeting we talked about making license optional. The main use cases being forwards compatibility and unlicensed e.g. internal software. Do you think those justify making the license field optional?

**Reply by lann** [reply]
I think it should be optional in the specification and probably required by any public-good registry.

**Comment by silesmo** [comment]
### Metadata

To summarize what we have discussed above. 

Place the content in a WebAssembly custom section called `registry-metadata` that would contain the following information in JSON:

- authors: List
- Description: Markdown blob
- License: SPDX
- Links: Key, Value
    - Where we have predefined meanings for the following keys:
    - documentation - link to documentation
    - homepage - link to projects homepage
    - repository - link to projects source repository
    - funding - link for users of the component to donate to the project
- Categories: List of strings

Do we want to restrict the categories somehow to not allow anything to be put into the category?

Do we want to use any compression algorithm on the metadata? If so which one Gzip, Brotli, zstd, LZ4?

I was planning on creating a library that can read and add the registry metadata to the wasm-tools repository similar to what has been done with the wasm-metadata crate. Is this what we want?

**Reply by lann** [reply]
> Do we want to use any compression algorithm on the metadata?

[per late discussion in meeting] In general I think its better to leave this up to the transport, i.e. normal HTTP negotiation.

**Reply by lann** [reply]
> Do we want to restrict the categories somehow to not allow anything to be put into the category?

I would think that this would be a per-registry policy, though (like for Links) we can certainly propose a default set.

**Reply by lann** [reply]
> I was planning on creating a library that can read and add the registry metadata to the wasm-tools repository similar to what has been done with the wasm-metadata crate. Is this what we want?

I think the rough consensus from last week's meeting was that this doesn't belong in `wasm-tools` (yet, at least). A separate crate in the registry repo would make sense to me.

**Reply by silesmo** [reply]
> > Do we want to use any compression algorithm on the metadata?
> 
> [per late discussion in meeting] In general I think its better to leave this up to the transport, i.e. normal HTTP negotiation.

Yes I agree. Just wanted to validate that this was what everyone else was thinking as well.
